horton/gbasis/ints.cpp

```diff
@@ -259,6 +263,113 @@ void GB2NuclearAttractionIntegral::add(double coeff, double alpha0, double alpha
 
 /*
 
+ GB2MomentIntegral
+
+ */
+
+
+double moment_helper(long n0, long n1, long n2, double pa, double pb, double pc, double gamma_inv) {
+    /*
+     The Obara Saika Scheme, equations A7 and A8 in "Efficient recursive computation of
+     molecular integrals over Cartesian Gaussian functions", S. Obara, A. Saika, Journal
+     of Chemical Physics, vol. 84, p. 3963, y. 1986.
+
+     Note that we use a shift in the indices:
+     (0_A|R(0)|0_B) in the paper is stored in work_mm[1][1][1].
+     In this allows to set all terms in with the angular momentum index becomes -1 to 0,
+     avoiding a lot of if-statements.
+    */
+
+    double result;
+    // if n2 == 0, we just need the overlap.
+    if (n2 == 0) {
+        result = gb_overlap_int1d(n0, n1, pa, pb, (2.0e0 * gamma_inv));
+    } else {
+        long m, l, n;
+        double work_mm[n0+2][n1+2][n2+2];
+
+        for (l=0; l < (n0+2); l++) {
+            for (m=0; m < (n1+2); m++) {
+                for (n=0; n < (n2+2); n++) {
+                    work_mm[l][m][n] = 0.0e0;
+                }
+            }
+        }
+
+        // The auxiliary overlap (0_A|R(0)|0_B) is stored in work_mm[1][1][1]:
+        work_mm[1][1][1] = gb_overlap_int1d(0, 0, pa, pb, (2.0e0 * gamma_inv));
+
+        // Equation A8 in the Obara-Saika paper for (0_A|R(mu + 1)|0_B):
+        for (n=1; n < (n2+1); n++) {
+            work_mm[1][1][n+1] = pc * work_mm[1][1][n] + (gamma_inv * (n-1) * work_mm[1][1][n-1]);
+        }
+
+        // Equation A7 in the Obara-Saika paper for (0_A|R(mu)|b + 1):
+        for (m=1; m < (n1+1); m++) {
+            for (n=1; n <= (n2+1); n++) {
+                work_mm[1][m+1][n] = pb * work_mm[1][m][n]
+                                                + (gamma_inv * (m-1) * work_mm[1][m-1][n])
+                                                + (gamma_inv * (n-1) * work_mm[1][m][n-1]);
+            }
+        }
+
+        // Equation A7 in the Obara-Saika paper for (a + 1|R(mu)|b):
+        for (l=1; l < (n0+1); l++) {
+            for (m=1; m <= (n1+1); m++) {
+                for (n=1; n <= (n2+1); n++) {
+                    work_mm[l+1][m][n] = pa * work_mm[l][m][n]
+                                                + (gamma_inv * (l-1) * work_mm[l-1][m][n])
+                                                + (gamma_inv * (m-1) * work_mm[l][m-1][n])
+                                                + (gamma_inv * (n-1) * work_mm[l][m][n-1]);
+                }
+            }
+        }
+
+        result = work_mm[n0+1][n1+1][n2+1];
+    }
+
+    return result;
+}
+
+
+GB2MomentIntegral::GB2MomentIntegral(long max_shell_type, long* xyz, double* center)
+: GB2Integral(max_shell_type), xyz(xyz) {
+    if (xyz[0] + xyz[1] + xyz[2] < 0)
+    throw std::domain_error(" sum < 0");
+}
```

Comment:
Is it only the sum that has to be positive. Shouldn't all components be positive?